### PR TITLE
Start sharing code between ILLink/ILCompiler test infra

### PR DESCRIPTION
**Pull request by sbomer**
This creates a new directory `Mono.Linker.Tests.Shared` with sources that are shared between the test infra of ILLink/ILCompiler. This starts by sharing most of the code in `TestRunner.cs`, renaming some types to make the sharing easier.

### CODE REVIEW

File: src/coreclr/tools/aot/ILCompiler.Tests/TestCasesRunner/TrimmingArgumentBuilder.cs [hunk 1]
```diff
@@ -8,20 +8,19 @@
...
 namespace Mono.Linker.Tests.TestCasesRunner
 {
-	public class ILCompilerOptionsBuilder
+	public class TrimmingArgumentBuilder
 	{
-		//public TrimmerOptions Options { get; } = new();
 		private readonly TestCaseMetadataProvider _metadataProvider;
 
-		public readonly ILCompilerOptions Options;
+		private readonly ILCompilerOptions _options;
```
Comment (agocke) [inline src/coreclr/tools/aot/ILCompiler.Tests/TestCasesRunner/TrimmingArgumentBuilder.cs:15]: Should the `_options` be zeroed out once you hit `Build()`?

**Review by agocke — APPROVED** [review]
LGTM, minor comments

**Comment by mrvoorhe** [comment]
Thanks.  This looks okay.

File: eng/Subsets.props [hunk 2]
```diff
@@ -352,7 +352,7 @@
       Test="true" Category="clr" Condition="'$(DotNetBuildFromSource)' != 'true'"/>
     <ProjectToBuild Include="$(CoreClrProjectRoot)tools\aot\ILCompiler.Compiler.Tests\ILCompiler.Compiler.Tests.csproj"
       Test="true" Category="clr" Condition="'$(DotNetBuildFromSource)' != 'true' and '$(NativeAotSupported)' == 'true'"/>
-    <ProjectToBuild Include="$(CoreClrProjectRoot)tools\aot\Mono.Linker.Tests\Mono.Linker.Tests.csproj"
+    <ProjectToBuild Include="$(CoreClrProjectRoot)tools\aot\ILCompiler.Tests\ILCompiler.Tests.csproj"
```
Comment (vitek-karas) [inline eng/Subsets.props:355]: I would prefer different name. `ILCompiler.Tests` sounds like "This is the main test assembly for ILCompiler", which is not true. We could probably use `ILCompiler.Trim.Tests` as vast majority of the tests are around trimming, some are around AOT diagnostics, but that's a minority.

File: src/tools/illink/test/Mono.Linker.Tests.Shared/Mono.Linker.Tests.Shared.projitems [hunk 3]
```diff
@@ -0,0 +1,5 @@
+<Project>
```
Comment (vitek-karas) [inline src/tools/illink/test/Mono.Linker.Tests.Shared/Mono.Linker.Tests.Shared.projitems:1]: Personally I would rename this to something like `Trim.Tests.Shared.projitems` (and the directory and so on).
Eventually we should also rename `ILLink.Shared` to something else - maybe `Trim.Shared`, or `Trimming.Shared`. It's weird to have ILLink in that name for several reasons :-).

File: src/tools/illink/test/Mono.Linker.Tests/TestCasesRunner/TrimmingDriver.cs [hunk 4]
```diff
@@ -5,13 +5,13 @@
 
 namespace Mono.Linker.Tests.TestCasesRunner
 {
-	public class LinkerDriver
+	public class TrimmingDriver
```
Comment (vitek-karas) [inline src/tools/illink/test/Mono.Linker.Tests/TestCasesRunner/TrimmingDriver.cs:8]: Even more reason to call the shared stuff `Trimming.Shared`/`Trimming.Tests.Shared` ;-).

File: src/tools/illink/test/Mono.Linker.Tests.Shared/TestRunner.cs [hunk 5]
```diff
@@ -99,41 +124,50 @@ private ManagedCompilationResult Compile (TestCaseSandbox sandbox, TestCaseCompi
...
-
-			var builder = _factory.CreateTrimmerOptionsBuilder (metadataProvider);
-
-			AddLinkOptions (sandbox, compilationResult, builder, metadataProvider);
-
-			var logWriter = new TestLogWriter ();
-			var trimmingResults = trimmer.Trim (builder.Options, logWriter);
-
-			return new ILCompilerTestCaseResult (testCase, compilationResult.InputAssemblyPath, compilationResult.ExpectationsAssemblyPath, sandbox, metadataProvider, compilationResult, trimmingResults, logWriter);
+			var trimmingCustomizations = CustomizeLinker (trimmer, metadataProvider);
```
Comment (vitek-karas) [inline src/tools/illink/test/Mono.Linker.Tests.Shared/TestRunner.cs:136]: This should be called `CustomizeTrimmer`.

**Review by vitek-karas — COMMENTED** [review]
Sorry for the late response. Only naming comments - suggestions.

**Pull request merged**